### ISSUE TITLE
Type inference too eager to evaluate at compile type leads to compiler error

### ISSUE THREAD

**Issue by fawzi** [ziglang/zig#13954]
### Zig Version

0.10.0 , 0.11.0-dev.814+077b003c5

### Steps to Reproduce and Observed Behavior

executing zig test bug.zig on the attached file [bug.zig.txt](https://github.com/ziglang/zig/files/10236567/bug.zig.txt) one gets

```
bug.zig:28:20: error: value with comptime-only type 'comptime_int' depends on runtime control flow
  const myPoints = switch (self.myPlay) {
                   ^~~~~~
bug.zig:28:32: note: runtime control flow here
  const myPoints = switch (self.myPlay) {
                           ~~~~^~~~~~~
```

This can be fixed by adding an explicit type declaration `:i32`to the variable.
Initially I had this trying to directly return the expression, but then simplified it to the current example.

I just tested with the latest zig and the bug is still present.

### Expected Behavior

compile and pass the test without errors

**Comment by mlugg** [comment]
The error is expected in this case. If an int type isn't explicitly specified, it's always assumed to be `comptime_int` by design. However, a direct `return` of this expression probably should work, as it should forward the return type through the expression.

**Comment by fawzi** [comment]
ok I did not know that, it was unexpected to me. I checked the single expression at least with 0.11.0-dev.814+077b003c5 worked, but something like

`		return switch (self.myPlay) {
			TickTackToe.Rock => 1,
			TickTackToe.Paper => 2,
			TickTackToe.Scissors => 3,
		} + switch (self.myPlay) {
			TickTackToe.Rock => 2,
			TickTackToe.Paper => 1,
			TickTackToe.Scissors => 3,
		} ;`

fails with
`fawzi@MacBook-Air-di-Fawzi 2 % zig test bug.zig
bug.zig:22:10: error: value with comptime-only type 'comptime_int' depends on runtime control flow
  return switch (self.myPlay) {
         ^~~~~~
bug.zig:22:22: note: runtime control flow here
  return switch (self.myPlay) {
                 ~~~~^~~~~~~`
 (this is basically equivalent to the expression that failed initially. So it seems that the + of two switch breaks the propagation of the type, and it then falls back to comptime_int. Anyway it was an error that confused me.
